presto-hive/src/main/java/com/facebook/presto/hive/rcfile/RcFilePageSourceFactory.java

```diff
@@ -67,66 +90,109 @@ public RcFilePageSourceFactory(TypeManager typeManager)
             TupleDomain<HiveColumnHandle> effectivePredicate,
             DateTimeZone hiveStorageTimeZone)
     {
-        // todo remove this when GC issues are resolved
-        if (true) {
+        if (!isRcfileOptimizedReaderEnabled(session)) {
             return Optional.empty();
         }
 
+        RcFileEncoding rcFileEncoding;
         String deserializerClassName = getDeserializerClassName(schema);
-
-        RcFileBlockLoader blockLoader;
         if (deserializerClassName.equals(LazyBinaryColumnarSerDe.class.getName())) {
-            blockLoader = new RcBinaryBlockLoader();
+            rcFileEncoding = new BinaryRcFileEncoding();
         }
         else if (deserializerClassName.equals(ColumnarSerDe.class.getName())) {
-            blockLoader = new RcTextBlockLoader(hiveStorageTimeZone);
+            rcFileEncoding = createTextVectorEncoding(schema, hiveStorageTimeZone);
         }
         else {
             return Optional.empty();
         }
 
-        // determine which hive columns we will read
-        List<HiveColumnHandle> readColumns = ImmutableList.copyOf(filter(columns, not(HiveColumnHandle::isPartitionKey)));
-        List<Integer> readHiveColumnIndexes = ImmutableList.copyOf(transform(readColumns, HiveColumnHandle::getHiveColumnIndex));
-
-        // Tell hive the columns we would like to read, this lets hive optimize reading column oriented files
-        setReadColumns(configuration, readHiveColumnIndexes);
-
-        // propagate serialization configuration to getRecordReader
-        for (String name : schema.stringPropertyNames()) {
-            if (name.startsWith("serialization.")) {
-                configuration.set(name, schema.getProperty(name));
-            }
-        }
-
-        RCFile.Reader recordReader;
+        long size;
+        FSDataInputStream inputStream;
         try {
             FileSystem fileSystem = path.getFileSystem(configuration);
-            recordReader = new RCFile.Reader(fileSystem, path, configuration);
+            size = fileSystem.getFileStatus(path).getLen();
+            inputStream = fileSystem.open(path);
```

Comment:
Depending on assignment to `inputStream` being the last line of a `try` seems very fragile to me. It can easily be broken by future code changes. I think it's better to merge this `try` with the next.